tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,306 @@
+# Lint as: python3
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Baron-Adesi Whaley approximation of the Black Scholes equation to calculate
+ the price of a batch of American options."""
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+from tf_quant_finance.black_scholes import implied_vol_newton_root
+from tf_quant_finance.black_scholes.vanilla_prices import option_price as vanilla_prices
+from tf_quant_finance.math.gradient import value_and_gradient
+
+
+def adesi_whaley(*,
+                 volatilities,
+                 strikes,
+                 expiries,
+                 discount_rates,
+                 continuous_dividends=None,
+                 cost_of_carries=None,
+                 spots=None,
+                 forwards=None,
+                 is_call_options=None,
+                 dtype=None,
+                 max_iterations=20,
+                 tolerance=1e-8,
+                 name="american_price"):
+  """Computes the price for a batch of call or put options, using an approximate
+  pricing formula, the Baron-Adesi Whaley approximation
```

Comment:
I just noticed, could you please make the first line in a docstring fit one line:
http://google.github.io/styleguide/pyguide.html#382-modules

 Here mention this is for American options, something like
"""Computes prices of American options using the Baron-Adesi Whaley  approximation."""